File: 3. Minimum Number of Coins/Experiments/Minimum Elements.java

// Using Recursion 



public class Solution {
    
    
    static int solveRec(int[]num , int x){
        if (x==0)
            return 0;
        if (x<0)
            return Integer.MAX_VALUE;
        int mini = Integer.MAX_VALUE;
        for(int i = 0; i < num.length; i++){
            int ans = solveRec(num, x- num[i]);
            if (ans != Integer.MAX_VALUE)
                mini = Math.min(mini,1+ans);
        }
        return mini;
        
    }
    
    
    public static int minimumElements(int num[], int x) {
         int ans = solveRec(num,x);
//             if (ans == Integer.MAX_VALUE)
//             return -1;
//             else 
//             return ans;
         return ans == Integer.MAX_VALUE ?-1:ans;

    }
      
}




___________________________________________________________________

//Using Memorization / Top down



import java.util.*;
public class Solution {
    
    
    static int solveMem(int[]num , int x , int[]dp){
      
        if (x==0)
            return 0;
        if (x<0)
           return Integer.MAX_VALUE;
        
   
            if (dp[x] != -1)
          return dp[x];
       
        int mini = Integer.MAX_VALUE;;
        
        for(int i = 0; i < num.length; i++){
            int ans = solveMem(num, x- num[i], dp);
            if (ans != Integer.MAX_VALUE)
                mini = Math.min(mini,1+ans);
        }

    dp[x] = mini;
    
    return mini;
}
    
    
    public static int minimumElements(int num[], int x) {
         
         int dp [] = new int[x+1];
        Arrays.fill(dp, -1);
        
        int ans = solveMem(num , x , dp);
        // if (ans == Integer.MAX_VALUE)
        //     return -1;
        // else 
        //     return ans;
        return ans == Integer.MAX_VALUE ? -1:ans;//same as above 
     }
      
}


___________________________________________________________________

// Tabulation / Bottom -Up



import java.util.*;
public class Solution {
    
    
    static int solveTab(int[]num , int x ){
        int[] dp = new int[x + 1];
        Arrays.fill(dp,Integer.MAX_VALUE);
        dp[0] = 0;

        for (int i = 1; i <= x; i++) {
             for(int j = 0; j < num.length; ++j)//same as above
                  {if(i-num[j] >= 0 && dp[i - num[j]] != Integer.MAX_VALUE){
            dp[i] = Math.min(dp[i] , 1 + dp[i - num[j]]);
                      }
               }
           }
            
            if(dp[x] != Integer.MAX_VALUE) 
                return dp[x];
              else 
                return -1;
        
     }
         public static int minimumElements(int num[], int x) {
         return solveTab(num , x);
    }
}





//same modify simply 

import java.util.*;
public class Solution {
    public static int minimumElements(int num[], int x) {
        
    int[] dp = new int[x + 1];
        Arrays.fill(dp, x + 1);
        dp[0] = 0;

        for (int i = 1; i <= x; i++) {
            
            //  for(int j = 0; j < num.length; ++j)//same as above
            //       {if(i-num[j] >= 0 && dp[i - num[j]] != Integer.MAX_VALUE){
            // dp[i] = Math.min(dp[i] , 1 + dp[i - num[j]]);
            //           }
            //    }

            for (int j : num) {
                if ((i - j) >= 0) {
                    dp[i] = Math.min(dp[i], 1 + dp[(i - j)]);
                }
            }
        }

        return dp[x] == x + 1 ? -1 : dp[x];
    }

}

File: 03. Minimum Number of Coins/Experiments/Minimum Elements.cpp

// Using Recursion 

int solveRec(vector<int>nums , int x)
{
//     Base Case
    if(x == 0) return 0;
    if(x < 0) return INT_MAX;
    
//     Recursive case
    
    int mini = INT_MAX;
    
    for(int i = 0; i < nums.size(); ++i)
    {
        int ans = solveRec(nums , x-nums[i]);
        if(ans != INT_MAX)
        {
            mini = min(mini , ans+1);
        }
    }
    return mini;
}
int minimumElements(vector<int> &num, int x)
{
   
    int ans = solveRec(num , x);
    if(ans == INT_MAX) return -1;
    else return ans;
    
}




___________________________________________________________________


//Using Memorization / Top down

int solveMem(vector<int> &num , int x , vector<int>&dp)
{

    if(x == 0) 
        return 0;
    if(x < 0) 
        return INT_MAX;
    
    if(dp[x] != -1) 
        return dp[x];

    
    int mini = INT_MAX;
    
    for(int i = 0; i < num.size(); ++i)
    {
        int ans = solveMem(num , x-num[i] , dp);
        if(ans != INT_MAX)
        mini = min(mini , ans+1);
        
    }

    dp[x] = mini;
    
    return mini;
}


int minimumElements(vector<int> &num, int x)
{
    vector<int>dp(x+1 , -1);
    int ans = solveMem(num , x , dp);
    if(ans == INT_MAX) 
        return -1;
    else
        return ans;
}


___________________________________________________________________

// Tabulation / Bottom -Up


int solveTab(vector<int>nums , int x)

{
    vector<int>dp(x+1 , INT_MAX);
    int mini = 0;
    dp[0] = 0;
    
    for(int i =1 ; i <= x; ++i)
    {
        for(int j = 0; j < nums.size(); ++j)
        {
            if(i-nums[j] >= 0 && dp[i - nums[j]] != INT_MAX)
            dp[i] = min(dp[i] , 1 + dp[i - nums[j]]);
        }
    }
    
    if(dp[x] != INT_MAX)return dp[x];
    return -1;
                                       
                                       
    
}

int minimumElements(vector<int> &num, int x)
{
    return solveTab(num , x);
}





